### ISSUE TITLE
reset() doesn't reset DALIGenericIterator

### ISSUE THREAD

**Issue by choosehappy** [NVIDIA/DALI#4656]
This pipeline works as expected the first epoch through, but resetting the iterator doesn't work and thus no subsequent epochs can be processed.

This is a minimal example, in an LMDB i have encoded jpegs and use a pytorch dataloader to assemble the batch

Any ideas?

```python
#--- read images from an LMDB which has an encoded jpeg stored
class DatasetLMDB(object):
    def __init__(self, fname):
        self.fname=fname

        self.env= lmdb.open(fname,map_size=1_000_000_000, max_dbs=1)

        dbkey='imgs'.encode()
        self.db=self.env.open_db(dbkey,integerkey=True)
        
        with self.env.begin(db=self.db,write=False) as txn:
            self.nitems=txn.stat(self.db)['entries']
        self.txn = None

    def __getitem__(self, index):
        self.txn = self.txn if self.txn else self.env.begin(db=self.db,write=False)
        dbi=dill.loads(self.txn.get(index.to_bytes(4,byteorder=sys.byteorder)))
        [imgencode,label,r,c]=dbi
        return imgencode,(np.asarray([0], dtype=np.int32))
        
    def __len__(self):
        return self.nitems


dataset=DatasetLMDB("M171030_2-HE.lmdb")

def collate_fn(batch):
    return tuple(map(list,zip(*batch)))

## - use a pytorch data loader and a collation function to get 
dataloader = torch.utils.data.DataLoader(dataset,batch_size=batch_size,num_workers=16, 
                                            shuffle=True, collate_fn=collate_fn)


#----
pipe = Pipeline(batch_size=batch_size, num_threads=16, device_id=0)
with pipe:
    jpegs,label = fn.external_source(source=dataloader, num_outputs=2, dtype=[types.UINT8,types.UINT8])
    decode = fn.decoders.image(jpegs, device="mixed")
    pipe.set_outputs(decode)
pipe.build()


pipeiter=DALIGenericIterator(pipe,['imgs']) 

#--- works as expected first time through
# %%time
for data in tqdm(pipeiter):
    continue

#340it [00:05, 66.70it/s]
#CPU times: user 10.7 s, sys: 4.39 s, total: 15.1 s
#Wall time: 5.1 s


#--- reset iterator
pipeiter.reset()


#-- doesn't reset - note 0 iterations and 2ms wall time
# %%time
for data in tqdm(pipeiter):
    continue

#0it [00:00, ?it/s]
#CPU times: user 4.46 ms, sys: 0 ns, total: 4.46 ms
#Wall time: 2.83 ms
```

**Comment by JanuszL** [comment]
Hi @choosehappy,

Thank you for reaching out. `pipeiter.reset()` doesn't affect the state of the callback/iterable passed to the `external_source` operator. So either it should reset its state automatically, or in your case, you can set `cycle='raise'` in the `external_source`  to let it restart provided input every time it raises StopIteration.

**Comment by choosehappy** [comment]
Yes! That totally did it, thank you!

**Comment by prefer-potato** [comment]
Thanks @JanuszL ,  Unfortunately, i also encountered the same problem. But I dont know how to raises StopIteration?，and by the way, i dont know the relationship between these two 'batch_size' in source and datapipe.
`class Archive_dataset(object):
    def __init__(self, lmdb_dir, split, batch_size):
        self.lmdb_dir = lmdb_dir
        self.length = info["archive"][split]
        self.env = None
        self.batch_size = batch_size

    def _init_db(self):
        self.env = lmdb.open(
            self.lmdb_dir,
            subdir=os.path.isdir(self.lmdb_dir),
            readonly=True,
            lock=False,
            readahead=False,
            meminit=False,
        )
        with self.env.begin(write=False) as txn:
            self.length = loads_pyarrow(txn.get(b"__len__"))
            self.keys = loads_pyarrow(txn.get(b"__keys__"))

    def __iter__(self):
        self.i = 0
        self.n = self.length
        return self

    def __next__(self):
        batch = []
        labels = []
        if self.env is None:
            self._init_db()
        env = self.env
        txn = env.begin(write=False)
        for _ in range(self.batch_size):
            byteflow = txn.get(self.keys[self.i])
            ref = loads_pyarrow(byteflow)
            ori_img = cv2.imdecode(
                np.frombuffer(ref["img"], np.uint8), cv2.IMREAD_COLOR
            )
            img = cv2.cvtColor(ori_img, cv2.COLOR_BGR2RGB)
            im = cp.asarray(img)
            class_label = ref["class_label"]
            batch.append(im.astype(cp.uint8))
            labels.append(cp.array([name2label[class_label]], dtype=np.uint8))
            self.i = self.i + 1 % self.n
            # self.i += 1 
            # if self.i >= self.n:
            #     raise('StopIteration')
            #     self.i = 0

        return (batch, labels)
batch_size = 2
eii_gpu = Archive_dataset(
    lmdb_dir="archive/lmdb/aaa_data.lmdb",
    split="aaa",
    batch_size=2,
)

pipe = Pipeline(batch_size=batch_size, num_threads=1, device_id=0)
with pipe:
    images, labels = fn.external_source(source = eii_gpu, num_outputs=2, device="gpu", dtype=types.UINT8)
    output = fn.crop_mirror_normalize(
        images,
        dtype=types.FLOAT,
        crop=(224, 224),
        mean=[0., 0., 0.],
        std=[1., 1., 1.],
        crop_pos_x=fn.random.uniform(range=(0, 1)),
        crop_pos_y=fn.random.uniform(range=(0, 1)))
    pipe.set_outputs(output, labels)
pipe.build()

dataloader = DALIClassificationIterator(
        pipe, last_batch_policy=LastBatchPolicy.FILL, auto_reset=False
    )

for data in tqdm(dataloader):
    image = data[0]['data']
    print('i')`
